### PR TITLE
basic implementation of walk_forward_split() and train_test_split()

### PR DESCRIPTION
**Pull request by mkos**
Fixes #.

### CODE REVIEW

File: darts/utils/model_selection.py [hunk 1]
```diff
@@ -0,0 +1,79 @@
+from typing import Sequence, Optional, Union
+from ..timeseries import TimeSeries
+
+
+def train_test_split(
+        data: Sequence[TimeSeries],
```
Comment (hrzn) [inline darts/utils/model_selection.py:6]: Could you implement support for `Union[TimeSeries, Sequence[TimeSeries]]` ?

File: darts/utils/model_selection.py [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+        test_size: Optional[Union[float, int]] = 0.25,
+        axis: Optional[int] = 0,
+        n: Optional[int] = 0,
+        horizon: Optional[int] = 0,
+        ) -> Sequence[Sequence[TimeSeries]]:
```
Comment (hrzn) [inline darts/utils/model_selection.py:11]: Should be `Tuple[Sequence[TimeSeries]]` ? Or actually even `Union[Tuple[TimeSeries], Tuple[Sequence[TimeSeries]]]` to support splitting simple `TimeSeries`.

File: darts/utils/model_selection.py [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            index = test_size
+
+        return data[:index], data[index:]
+
+    elif axis == 1:
+
+        if horizon == 0 or n == 0:
+            raise AttributeError("You need to provide non-zero `horizon` and `n` parameters when axis=1")
+
+        train_set = []
```
Comment (hrzn) [inline darts/utils/model_selection.py:57]: This assumes that the datasets fit into memory.
Another option to scale to bigger datasets here could be to build our own Sequence implementation. Something like:
```
class MyTimeSeriesSequence(Sequence):
    def __init__(data):
        self.data = data

    def __getitem__(idx: int):
        ts = self.data[idx]
        ts_length = len(ts)
        ...
        return ts[:train_end_index], ts[test_start_index:]

...

return MyTimeSeriesSequence(data)
```

In this example it would make the return type of the method `Sequence[Tuple]` instead of `Tuple[Sequence]`, but it can be changed if `Tuple[Sequence]` makes more sense (wdyt about this?)

File: darts/utils/model_selection.py [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+
+            test_start_index = train_end_index - n
+
+            train_set.append(ts[:train_end_index])
+            test_set.append(ts[test_start_index:])
+
+        return train_set, test_set
+
+    else:
+        raise AttributeError()
```
Comment (hrzn) [inline darts/utils/model_selection.py:79]: Maybe an error message explaining what `axis` should be?

File: darts/tests/utils/test_model_selection.py [hunk 5]
```diff
@@ -0,0 +1,104 @@
...
+    def test_parameters_for_axis_0(self):
+        train_test_split(make_dataset(2, 10), axis=0, test_size=1)
+
+        # expecting no exception
+        self.assertTrue(True)
+
+    # test 2
+    # test parameteres for axis = 1
+    def test_parameters_for_axis_1_no_n(self):
+        with self.assertRaises(AttributeError):
```
Comment (hrzn) [inline darts/tests/utils/test_model_selection.py:29]: why should this fail?

File: darts/tests/utils/test_model_selection.py [hunk 6]
```diff
@@ -0,0 +1,104 @@
...
+            train_test_split(make_dataset(1, 10), axis=1, horizon=1)
+
+    def test_parameters_for_axis_1_no_horizon(self):
+        with self.assertRaises(AttributeError):
```
Comment (hrzn) [inline darts/tests/utils/test_model_selection.py:33]: why should this fail? you have a default value for horizon.

File: darts/tests/utils/test_model_selection.py [hunk 7]
```diff
@@ -0,0 +1,104 @@
...
+    # test 10
+    # test_split absolute number vertical split
+    # input is 4 timeseries of length 10, axis = 1, test_split = 2
+    def test_test_split_absolute_number_vertical(self):
+        train_set, test_set = train_test_split(make_dataset(4, 10), axis=1, test_size=2, n=1, horizon=2)
+
+        self.assertTrue(
+            verify_shape(train_set, 4, 8) and
+            verify_shape(train_set, 4, 4)
+        )
```
Comment (hrzn) [inline darts/tests/utils/test_model_selection.py:104]: Might be nice to add a test case for checking the train / val shapes when the timeseries have unequal lengths.

**Review by hrzn — COMMENTED** [review]
You still have some failing tests.

File: darts/utils/model_selection.py (hunk 1 above)
Comment (mkos) [reply darts/utils/model_selection.py:6]: done

File: darts/tests/utils/test_model_selection.py (hunk 5 above)
Comment (mkos) [reply darts/tests/utils/test_model_selection.py:27]: Since `n` and `horizon` are only valid parameters for `axis=1` they are set by default to 0 to cover the case of `axis=0`. But if the `axis=1` and any of these is set to 0, then it's an error.

File: darts/tests/utils/test_model_selection.py (hunk 6 above)
Comment (mkos) [reply darts/tests/utils/test_model_selection.py:31]: Since `n` and `horizon` are only valid parameters for `axis=1` they are set by default to 0 to cover the case of `axis=0`. But if the `axis=1` and any of these is set to 0, then it's an error.

File: darts/utils/model_selection.py (hunk 3 above)
Comment (mkos) [reply darts/utils/model_selection.py:59]: How about there's going to be additional parameter where you can provide your own implementation of sequence that defaults to `list()`?

File: darts/tests/utils/test_model_selection.py (hunk 7 above)
Comment (mkos) [reply darts/tests/utils/test_model_selection.py:82]: done.

File: darts/utils/model_selection.py (hunk 3 above)
Comment (mkos) [reply darts/utils/model_selection.py]: (I have implemented this behavior)

Comment (hrzn) [reply darts/utils/model_selection.py]: That would make it quite complex to use for users no? It means they would have to implement their own class. Whereas if you implement the class for them they don't have to worry about it. Is there a use-case where a class looking like the one I'm proposing above wouldn't work?

Comment (mkos) [reply darts/utils/model_selection.py]: Everything is fine with your implementation. I was always hesitant to implement things that may not be used at all, I'm a big fan of "less is more". I'm assuming, based on what you said about use cases, that in 90% of the cases `list()` will be completely sufficient. For the remaining 10% of the cases, where large datasets are used, I'm assuming these will be the power users that know how to implement it. I'm also a big proponent of flexibility with sane defaults. And I can imagine some exotic cases where custom implementation will be required (eg. storing data on the disk, using a joblib to manage memory or using tensors as an internal repr etc.).

File: darts/utils/model_selection.py [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+            ts_length = len(ts)
+            train_end_index = ts_length - horizon
+
+            if 0 < test_size < 1:
+                test_size = int((ts_length - horizon) * (test_size))
+                print('test_size: ', test_size)
+            if train_end_index < n:
+                raise AttributeError("Test size is too small: training timeseries is of 0 size")
+
+            test_start_index = ts_length - horizon - n - test_size - 1
```
[the remark this answers is not in the document]
Comment (mkos) [reply darts/utils/model_selection.py:93]: - add an option for type of vertical cut `simple` and `model-aware` (default)
- change `n` to `input_size`

**Comment by mkos** [comment]
Will open regular PR.
